process/executionTrack/executionResultsTracker.go

```diff
@@ -0,0 +1,165 @@
+	nonceHashes            *nonceHashes
+}
+
+// NewExecutionResultsTracker will create a new instance of *executionResultsTracker
+func NewExecutionResultsTracker() (*executionResultsTracker, error) {
+	return &executionResultsTracker{
+		executionResultsByHash: make(map[string]*block.ExecutionResult),
+		nonceHashes:            newNonceHashes(),
+	}, nil
+}
+
+// AddExecutionResult will add the provided execution result in tracker
+// It will return true if the execution result was added in the tracker
+func (est *executionResultsTracker) AddExecutionResult(executionResult *block.ExecutionResult) error {
+	if executionResult == nil {
+		return ErrNilExecutionResult
+	}
+
+	est.mutex.Lock()
+	defer est.mutex.Unlock()
+	if est.lastNotarizedResult == nil {
+		return ErrNilLastNotarizedExecutionResult
+	}
+
+	if est.lastNotarizedResult.Nonce >= executionResult.Nonce {
+		return fmt.Errorf("execution results nonce(%d) is lower than last notarized nonce(%d)", executionResult.Nonce, est.lastNotarizedResult.Nonce)
+	}
+
+	est.executionResultsByHash[string(executionResult.HeaderHash)] = executionResult
+	est.nonceHashes.addNonceHash(executionResult.Nonce, string(executionResult.HeaderHash))
+
+	return nil
+}
+
+// GetPendingExecutionResults will return the pending execution results
+func (est *executionResultsTracker) GetPendingExecutionResults() ([]*block.ExecutionResult, error) {
+	executionResults := make([]*block.ExecutionResult, 0)
+	est.mutex.Lock()
+	defer est.mutex.Unlock()
+
+	for _, executionResult := range est.executionResultsByHash {
+		executionResults = append(executionResults, executionResult)
+	}
+
+	if len(executionResults) == 0 {
+		return executionResults, nil
+	}
+
+	sort.Slice(executionResults, func(i, j int) bool {
+		return executionResults[i].Nonce < executionResults[j].Nonce
+	})
+
+	firstElementHasCorrectNonce := executionResults[0].Nonce == est.lastNotarizedResult.Nonce+1
+	if !firstElementHasCorrectNonce {
+		return nil, ErrDifferentNoncesConfirmedExecutionResults
+	}
+
+	for idx := 0; idx < len(executionResults)-1; idx++ {
+		hasConsecutiveNonces := executionResults[idx].Nonce+1 == executionResults[idx+1].Nonce
+		if !hasConsecutiveNonces {
+			return nil, ErrDifferentNoncesConfirmedExecutionResults
+		}
+	}
+
+	return executionResults, nil
+}
+
+// GetExecutionResultByHash will return the execution results by hash
+func (est *executionResultsTracker) GetExecutionResultByHash(hash []byte) (*block.ExecutionResult, error) {
+	est.mutex.RLock()
+	defer est.mutex.RUnlock()
+
+	result, found := est.executionResultsByHash[string(hash)]
+	if !found {
+		return nil, errors.New("cannot find execution results by provided hash	")
+	}
+
+	return result, nil
+}
+
+// GetExecutionResultsByNonce will return the execution results by nonce
+func (est *executionResultsTracker) GetExecutionResultsByNonce(nonce uint64) ([]*block.ExecutionResult, error) {
+	executionResults := make([]*block.ExecutionResult, 0)
```

Comment:
```suggestion
	executionResults := make([]*block.ExecutionResult, 0, len(hashes))
```
and move it under mutex protection

Reply:
fixed